From 9c96c3698c5cf578364966f0d88575ef0442e705 Mon Sep 17 00:00:00 2001
From: Jiebing Li <jiebing.li@intel.com>
Date: Fri, 1 Aug 2014 15:49:08 +0800
Subject: [PATCH 3/3] usb: dwc3-device: add OTG configuration for CHT platform

Intel Cherrytrail platform has it's own OTG PHY driver, so this patch
enables hooking of CHT OTG PHY driver in dwc3 driver, and add PCI id
for dwc3 controller on CHT platform.

Change-Id: Ide662ece5d876eea91c04a53d69af9bed806c2ba
Signed-off-by: geoffroy.weisenhorn <weisenhorn.geoffroy.@intel.com>
Signed-off-by: Jiebing Li <jiebing.li@intel.com>
Signed-off-by: Wu, Hao <hao.wu@intel.com>
---
 drivers/usb/dwc3/dwc3-pci.c |   12 +++++++++++-
 drivers/usb/dwc3/gadget.c   |    8 ++++++++
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/dwc3/dwc3-pci.c b/drivers/usb/dwc3/dwc3-pci.c
index 8325d3b180f7..01ff044ae4b9 100644
--- a/drivers/usb/dwc3/dwc3-pci.c
+++ b/drivers/usb/dwc3/dwc3-pci.c
@@ -32,6 +32,7 @@
 #define PCI_DEVICE_ID_SYNOPSYS_HAPSUSB3	0xabcd
 #define PCI_DEVICE_ID_INTEL_BYT		0x0f37
 #define PCI_DEVICE_ID_INTEL_MRFLD	0x119e
+#define PCI_DEVICE_ID_INTEL_CHT		0x22b7
 
 struct dwc3_pci {
 	struct device		*dev;
@@ -45,11 +46,19 @@ static int dwc3_pci_register_phys(struct dwc3_pci *glue)
 {
 	struct usb_phy_gen_xceiv_platform_data pdata;
 	struct platform_device	*pdev;
+	struct pci_dev	*pci_dev;
 	int			ret;
 
 	memset(&pdata, 0x00, sizeof(pdata));
 
-	pdev = platform_device_alloc("usb_phy_gen_xceiv", 0);
+	pci_dev = to_pci_dev(glue->dev);
+
+	if (pci_dev->vendor == PCI_VENDOR_ID_INTEL &&
+			pci_dev->device == PCI_DEVICE_ID_INTEL_CHT)
+		pdev = platform_device_alloc("intel-cht-otg", 0);
+	else
+		pdev = platform_device_alloc("usb_phy_gen_xceiv", 0);
+
 	if (!pdev)
 		return -ENOMEM;
 	pdev->dev.parent = glue->dev;
@@ -252,6 +261,7 @@ static const struct pci_device_id dwc3_pci_id_table[] = {
 	},
 	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_BYT), },
 	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_MRFLD), },
+	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_CHT), },
 	{  }	/* Terminating Entry */
 };
 MODULE_DEVICE_TABLE(pci, dwc3_pci_id_table);
diff --git a/drivers/usb/dwc3/gadget.c b/drivers/usb/dwc3/gadget.c
index ac97025b0e2e..c118574743f3 100644
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -1543,6 +1543,14 @@ static int dwc3_gadget_start(struct usb_gadget *g,
 
 	pm_runtime_get_sync(dwc->dev);
 
+	if (dwc->usb2_phy) {
+		ret = otg_set_peripheral(dwc->usb2_phy->otg, &dwc->gadget);
+		if (ret == -ENOTSUPP)
+			dev_info(dwc->dev, "no OTG driver registered\n");
+		else if (ret)
+			goto err0;
+	}
+
 	irq = platform_get_irq(to_platform_device(dwc->dev), 0);
 	ret = request_threaded_irq(irq, dwc3_interrupt, dwc3_thread_interrupt,
 			IRQF_SHARED, "dwc3", dwc);
-- 
1.7.9.5

